src/terminal-screen.c

```diff
@@ -1761,24 +1761,46 @@ terminal_screen_button_press (GtkWidget      *widget,
 		}
 	}
 
-	// right button with no Ctrl, Alt or Shift
-	if (event->button == 3 &&
-	        (state & (GDK_SHIFT_MASK | GDK_CONTROL_MASK | GDK_MOD1_MASK)) == 0)
+	// right button with Shift, no Ctrl or Alt
+	if (event->button == 3)
 	{
-		TerminalScreenPopupInfo *info;
+		if (!(state & (GDK_SHIFT_MASK | GDK_CONTROL_MASK | GDK_MOD1_MASK)))
+		{
+			if (button_press_event && button_press_event(widget, event))
```

Comment:
Since this is a backport, adding the same comments that `gnome-terminal` has is very useful to understanding what's going on:
```
          /* on right-click, we should first try to send the mouse event to
           * the client, and popup only if that's not handled. */
```